drv/sidecar-mainboard-controller/src/tofino2.rs

```diff
@@ -231,3 +232,429 @@ impl Sequencer {
+    Bar0 = 0,
+    Msi = 1 << 28,
+    Cfg = 2 << 28,
+}
+
+/// A few of the Tofino registers which are used in code below. These are found
+/// in 631384-0001_TF2-Top-Level_Register_Map_05062021.html as provided by
+/// Intel.
+#[derive(Copy, Clone, PartialEq, FromPrimitive, AsBytes)]
+#[repr(u32)]
+pub enum TofinoRegisters {
+    Scratchpad = 0x0,
+    FreeRunningCounter = 0x10,
+    SpiOutData0 = (0x80000 | 0x120),
+    SpiOutData1 = (0x80000 | 0x124),
+    SpiInData = (0x80000 | 0x12c),
+    SpiCommand = (0x80000 | 0x128),
+    SpiIdCode = (0x80000 | 0x130),
+}
+
+impl From<TofinoRegisters> for u32 {
+    fn from(r: TofinoRegisters) -> Self {
+        r as u32
+    }
+}
+
+/// SPI EEPROM instructions, as per for example
+/// https://octopart.com/datasheet/cat25512vi-gt3-onsemi-22302617.
+#[derive(Copy, Clone, PartialEq, FromPrimitive, AsBytes)]
+#[repr(u8)]
+pub enum SpiEepromInstruction {
+    // WREN, enable write operations
+    WriteEnable = 0x6,
+    // WRDI, disable write operations
+    WriteDisable = 0x4,
+    // RDSR, read Status register
+    ReadStatusRegister = 0x5,
+    // WRSR, write Status register. See datasheet for which bits can actually be
+    // written.
+    WriteStatusRegister = 0x1,
+    // READ, read a number of bytes from the EEPROM
+    Read = 0x3,
+    // WRITE, write a number of bytes to the EEPROM
+    Write = 0x2,
+}
+
+impl From<SpiEepromInstruction> for u8 {
+    fn from(i: SpiEepromInstruction) -> Self {
+        i as u8
+    }
+}
+
+pub struct DebugPort {
+    fpga: FpgaUserDesign,
+}
+
+impl DebugPort {
+    pub fn new(task_id: userlib::TaskId) -> Self {
+        Self {
+            fpga: FpgaUserDesign::new(
+                task_id,
+                MainboardController::DEVICE_INDEX,
+            ),
+        }
+    }
+
+    pub fn state(&self) -> Result<DebugPortState, FpgaError> {
+        self.fpga.read(Addr::TOFINO_DEBUG_PORT_STATE)
+    }
+
+    pub fn set_state(&self, state: DebugPortState) -> Result<(), FpgaError> {
+        self.fpga
+            .write(WriteOp::Write, Addr::TOFINO_DEBUG_PORT_STATE, state)
+    }
+
+    pub fn read_direct(
+        &self,
+        segment: DirectBarSegment,
+        offset: impl Into<u32>,
+    ) -> Result<u32, FpgaError> {
+        let state = self.state()?;
+
+        if !state.send_buffer_empty() || !state.receive_buffer_empty() {
+            return Err(FpgaError::InvalidState);
+        }
+
+        // Add the segement base address to the given read offset.
+        let address = segment as u32 | offset.into();
+
+        // Write the opcode.
+        self.fpga.write(
+            WriteOp::Write,
+            Addr::TOFINO_DEBUG_PORT_BUFFER,
+            u8::from(DebugRequestOpcode::DirectRead),
+        )?;
+
+        // Write the address.
+        for b in address.as_bytes().iter() {
+            self.fpga.write(
+                WriteOp::Write,
+                Addr::TOFINO_DEBUG_PORT_BUFFER,
+                *b,
+            )?;
+        }
```

Comment:
`u32` already implements `as_bytes`, so I think we can avoid an explicit loop:
```suggestion
        // Write the address.
        self.fpga.write(
            WriteOp::Write,
            Addr::TOFINO_DEBUG_PORT_BUFFER,
            address,
        )?;
```

Reply:
I've added comments explaining why this is done in a loop. Similar for all the cases below.